### DISCUSSION TITLE
Discrepancy between the Mopeka Add On UI and the conf files

### DISCUSSION THREAD

**Discussion by ethanlchristensen** [jgyates/genmon#1122] [Questions] [answered]
### Checklist

- [X] I made sure that there are *no existing issues or discussions* - [open issues](https://github.com/jgyates/genmon/issues),[closed issues](https://github.com/jgyates/genmon/issues?q=is%3Aissue+is%3Aclosed),[open discussion](https://github.com/jgyates/genmon/discussions), [closed discussion](https://github.com/jgyates/genmon/discussions/categories/questions?discussions_q=is%3Aclosed) - which may answer my question or that I could contribute my information to.
- [X] I have read the [Known Issues Wiki Page](https://github.com/jgyates/genmon/wiki/Appendix-D-Known-Issues) and my question / issue is not listed.
- [X] I have provided the version of the software and the operating system in my post.
- [X] If I have an issue with genmon not communicating with the controller, I have read and followed the [serial troubleshooting guide](https://github.com/jgyates/genmon/wiki/3.6---Serial-Troubleshooting).

### Version

1.19.03

### Have you submitted your log files from the About page? Note that outbound email must be working for logs to be submitted.

I have submitted logs

### What is/are your question(s)?

I was curious if anyone knew what would cause a discrepancy between the Add On UI and the actual conf files?

I am trying to get the Mopeka Add On to work, but looking that the log files, I do not see a Mopeka log file, but have enabled the add on.

I checked the Mopeka conf file and it did not have the MAC address or tank size I entered in the add on UI. I added these manually to the conf file and updated the genloader.conf to enable to plugin (it was disabled despite being turned on in the UI).

### Additional information

![image](https://github.com/jgyates/genmon/assets/55725575/60e3f1b1-db89-4727-9e9e-f9e7f8ee7193)
![image](https://github.com/jgyates/genmon/assets/55725575/a1ca814f-29af-454f-818e-6d3bdaa8d950)
![image](https://github.com/jgyates/genmon/assets/55725575/3bcbcdec-bb9b-48f6-a2a9-89be250bc5df)

**Comment by jgyates** [comment, marked as answer]
Hi @ethanlchristensen 

The log files for all genmon components are stored in /var/log.
The working copy of config files for all genmon components are stored /etc/genmon
The config files located in the project directory (~/genmon/conf) are the orginal config files. These get updated when you do a software update so the working copy in /etc/genmon is then one that the UI modifies.

Let me know if you have any other questions.

**Reply by ethanlchristensen** [reply]
Would the absence of the genmopeka log indicate that this did not start up?

**Reply by jgyates** [reply]
yes. When you installed genmon, did you follow the steps in the wiki? I am guessing that genmopeka.py is not loading due to missing python libraries. The required python libraries are installed when you run ./genmonmaint.sh -i

https://github.com/jgyates/genmon/wiki/3.3--Setup-genmon-software

If you don't use the git to download genmon, then the software update will not work. If you don't run the setup script, all of the required libraries will be installed. If the install failed due to internet failures then the libraries will fail to install.

You can run the script in the genmon folder now if you have not done so before. The script will ask a few questions. If you select Yes when it ask to copy the conf files, it will overwrite the files in the /etc/genmon folder. Otherwise, select N for No.

genloader.py (which is called with startgenmon.sh) will attempt to load most of the libraries that are missing, except some of the large ones like [fluids](https://pypi.org/project/fluids/). This library is only installed by the install script.
         
also, I did not receive you logs. Your outbound email must be setup and working to send the logs files form the About page.

**Reply by skipfire** [reply]
The preloaded installs are done with `./genmonmaint.sh -i -n` and the image was last made fresh in March.

**Reply by ethanlchristensen** [reply]
Yes I did do this. I noticed when running genmopeka.py just now it says bleson is missing, but when I pip freeze (both globally and in the venv) bleson is installed. I suppose this is the issue then

**Comment by jgyates** [comment]
https://github.com/jgyates/genmon/wiki/Appendix-S---Working-in-a-Managed-Python-Environment

as a test you could load the virtual environment:
 
     cd genmon
     source ./genenv/bin/activat

Then load python3 in the environment:

   sudo ./genenv/bin/python3 

then try to bleson and themodules needed by belson in python3

     from enum import Enum
     from typing import Optional
     from bleson import get_provider, BDAddress
     from bleson.core.hci.type_converters import rssi_from_byte
     from bleson.core.hci.constants import GAP_MFG_DATA, GAP_NAME_COMPLETE

If any of these fail the genmopeka.py will print the error "The required library bleson is not installed."

 Exit python3
 

       exit()

Exit the virtual environment:

      deactivate

**Reply by ethanlchristensen** [reply]
Sure thing, here is the output from the little test

![image](https://github.com/jgyates/genmon/assets/55725575/df3ba0ad-c067-4b90-80e1-7670ac5e42c8)

**Reply by jgyates** [reply]
run the above again but with the python3 from the environemnt:

    sudo ./genenv/bin/python3

and also run genmopeka.py with the python3 from the add on

    sudo ./genenv/bin/python3 ./addon/genmopeka.py

**Reply by ethanlchristensen** [reply]
This seemed to have worked

![image](https://github.com/jgyates/genmon/assets/55725575/dc255fb9-3c44-42d1-ac1b-2c6cac31fed1)

**Reply by jgyates** [reply]
if you use the startgenmon.sh script then it will load the add on programs that are enabled via the UI. The startgenmon.sh script calls genloader.py, which loads genmon and any addons specified by /etc/genmon/genloader.conf. It will load then in the virtual environment using the environments python version.

There are three options with startgenmon.sh

./startgenmon.sh start
./startgenmon.sh stop
./startgenmon.sh restart

**Comment by ethanlchristensen** [comment]
Okay, running the stop --> then start, I do see a mopeka log file, but, yet to see the gauge pop up. I had the gauge working yesterday, but had to restart the set up due to some issues I was having.

**Reply by jgyates** [reply]
try refreshing your browser. 

If that does not work then restart your pi. The scrtip will only stop modules that have been started with genloader. If you start genmon outside of the script then it will not track the process ID.

**Comment by ethanlchristensen** [comment]
I have this setup to start genmon on boot. I am guessing this will activate the environment similar to these commands?
./startgenmon.sh start
./startgenmon.sh stop
./startgenmon.sh restart

**Reply by jgyates** [reply]
Yes, if you told the setup script (genmonmaint.sh) to start on boot, then it will add ./startgenmon.sh start to crontab, which will start on boot. The scripts in the genmon project (genmonmaint.sh and startgenmon.sh) are aware of virtual environments and handled that if virtual environments are enabled in the global python environment.

**Comment by ethanlchristensen** [comment]
In the genmon log, I noticed this line, would this indicate the VENV is not loaded?
![image](https://github.com/jgyates/genmon/assets/55725575/be2fe23d-4f20-4bfc-81af-9e470d14f347)

**Reply by jgyates** [reply]
yes.

Were you loading genmon with the script when these entries were created?

The raspberry pi OS (starting with bookworm) uses python 3.11 and has virtual environments enabled by default.

**Reply by jgyates** [reply]
to clarify, the log entry you are referencing will indicate if the python3 program is running in a virtual environment by querying the python runtime environment (i.e. is this python script executing in a virtual environment).

the scripts will query the OS / file system to see if virtual environment are required by the OS (they are with bookworm) and will create and use a virtual environments for use with genmon (if enforced by the OS).

If you manually loaded genmon without using the startgenmon.sh scrip, you will see VEnv:False as you are bypassing the virtual environments.

**Reply by ethanlchristensen** [reply]
The only way I have ever loaded up genmon is either with the start-on-boot cron job, or using the startgenmon.sh script.

I think I should also note that I disabled the enforcement of the virtual environments by moving the file:
`sudo mv /usr/lib/python3.11/EXTERNALLY-MANAGED /usr/lib/python3.11/EXTERNALLY-MANAGED.old`

Perhaps I shouldn't have done this. But, if the cron job activates the environment, I figured this wouldn't be an issue.

**Reply by ethanlchristensen** [reply]
I stopped genmon and started back up using `startgenmon.sh` and got the following
```bash
2024-05-25 15:13:42,700 : Generator Monitor Shutdown
2024-05-25 15:14:18,552 : Selected Controller: generac_evo_nexus
2024-05-25 15:14:18,841 : GenMon Loaded for site: Optic Road using python 3.11: VEnv: False
```

**Reply by jgyates** [reply]
ok. that makes sense. Renaming this file disables the managed environment requirement so the install scripts will install everything in the system python environment. 

The crontab boot entry also calls startgenmon.sh

The crontab entry to start on boot works with the environment requirement enabled.

**Reply by ethanlchristensen** [reply]
Okay, I set managedpackages to true and set the python command to the ENV in the startgenmon script and I can see the gauge now . . waiting for it to update.

```bash
genmondir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
pythoncommand="$genmondir/genenv/bin/python3"
pipcommand="pip3"
config_path=""
usepython3=true
found_action=false
managedpackages=true
. . . .
```

Is the polling rate for the mopeka sensor minutes or seconds. I noticed in the info it said polling rate in "minutes" but the default is 60 "seconds".
![image](https://github.com/jgyates/genmon/assets/55725575/4ff1b9c3-6d6d-496d-aecd-a22c60a8a723)

**Reply by jgyates** [reply]
I updated the comment in the tool tip bubble so you should see this if you update from the About page (which will restart the software). The software version is not updated so the version has not changed.

also, keep in mind that if you modify the source or scripts, they will be overwritten if you update the software from the About page. The "mamagedpackages" flag is updated by the script based on the presence of the /usr/lib/python3.11/EXTERNALLY-MANAGED  file.

**Reply by ethanlchristensen** [reply]
I am guessing then, the way to solve this problem I created it to create my own startup job? (or reinstall)

**Reply by ethanlchristensen** [reply]
I don't mind manually updating the startup script, the only thing I noticed is if I change an addon setting, it seems like it reverts the changes and genmon restarts with venv false.

```bash
#### starting up genmon after manual edits ####
2024-05-25 16:11:48,597 : Generator Monitor Shutdown
2024-05-25 16:11:59,791 : Selected Controller: generac_evo_nexus
2024-05-25 16:12:00,092 : GenMon Loaded for site: Optic Road using python 3.11: VEnv: True

#### Making a change to an addon setting ####
2024-05-25 16:14:22,397 : Generator Monitor Shutdown
2024-05-25 16:14:25,590 : Selected Controller: generac_evo_nexus
2024-05-25 16:14:25,950 : GenMon Loaded for site: Optic Road using python 3.11: VEnv: False
```

**Reply by jgyates** [reply]
If you restored the /usr/lib/python3.11/EXTERNALLY-MANAGED file, then just run the install script again to install the libraries in the virtual environment. This will allow genmon to run in the virtual environment. This will still have some of the libraries installed in the system python environment. My guess is that this will likely be OK. Virtual environments exist mostly to protect apps. The default python environment is pretty basic on Raspbian.

Otherwise, just keep the /usr/lib/python3.11/EXTERNALLY-MANAGED file modified and the scripts will use the system python environment. 

Either one will work, you just need to know how it was installed if you want to bypass the startgenmon.sh script like we did above (for testing purposes). You mentioned that you were running bookworm so I assumed you had not edited the /usr/lib/python3.11/EXTERNALLY-MANAGED file.

**Reply by ethanlchristensen** [reply]
Okay, final post . . .
I just restored the EXTERNALLY-MANAGED in the /usr/lib/python3.11 directory, and now all works, even when I change addon settings. So I guess we are also good for updating genmon too as the script will recognize this EXTERNALLY-MANAGED file now.

Thanks for the responses @jgyates @skipfire 

![image](https://github.com/jgyates/genmon/assets/55725575/a89f1729-0419-4f18-a671-bb9c06839007)

![image](https://github.com/jgyates/genmon/assets/55725575/69638bb7-ad85-4315-aba1-e38fa34815e8)

**Reply by jgyates** [reply]
Great, I am glad it is working for you.
